core/src/workflow/workflow_tasks/concurrency_manager.rs

```diff
@@ -266,15 +266,35 @@ impl WorkflowConcurrencyManager {
         F: for<'a> FnOnce(&'a mut WorkflowManager) -> BoxFuture<Result<Fout>>,
         Fout: Send + Debug,
     {
-        let readlock = self.runs.read();
-        let m = readlock
-            .get(run_id)
-            .ok_or_else(|| WFMachinesError::Fatal("Missing workflow machines".to_string()))?;
-        // This holds a non-async mutex across an await point which is technically a no-no, but
-        // we never access the machines for the same run simultaneously anyway. This should all
-        // get fixed with a generally different approach which moves the runs inside workers.
-        let mut wfm_mutex = m.wfm.lock();
-        let res = mutator(&mut wfm_mutex).await;
+        // TODO: This is less than ideal. We must avoid holding the read lock on the overall machine
+        //  map while async-ly mutating the inner machine. So, we take it out of the map and put
+        //  it back when we're done. We should restructure things to avoid the top-level lock
+        //  on the map.
+
+        let mut wfm = {
+            let readlock = self.runs.read();
+            let m = readlock
+                .get(run_id)
+                .ok_or_else(|| WFMachinesError::Fatal("Missing workflow machines".to_string()))?;
+            let x = m
+                .wfm
+                .lock()
+                .take()
+                .expect("WorkflowManager inside a ManagedRun is never None");
+            x
+        };
+
+        let res = mutator(&mut wfm).await;
+
+        {
+            let readlock = self.runs.read();
+            // SAFETY: Since runs are accessed serially the machines cannot go missing while
+            // we are modifying them for a given run
+            let m = readlock
+                .get(run_id)
+                .expect("Machine cannot be missing as each run is accessed serially");
```

Comment:
If this is the case, can you hold the `wfm.lock()` across the life of this call even though you don't hold the map lock?

Reply:
I cannot - and this was in fact the non-obvious part. The problem is that I have a _reference_ to the mutex through the map-level readlock. So if I keep the inner lock guard around, I am by definition _also_ keeping the outer map lock guard around since I don't own that mutex, I am only borrowing it.

Reply:
Adding the scoped blocks clarifies this since now you get lifetime problems if you try to only keep the inner mutex around but not the outer one.

Reply:
@cretz You made me realize this can look much nicer with an Arc though instead of this option nonsense, so thanks for mentioning.

Reply:
Makes sense you need the map in scope to access its value as a reference :+1:

I do think a future refactor that moves mutability/locking down to a lower level than this higher level "access" abstraction could make sense.